drivers/network/netio/netio.c

```diff
@@ -0,0 +1,882 @@
+				&nameUnicode, FILE_DEVICE_UNKNOWN,
+				FILE_DEVICE_SECURE_OPEN, FALSE, &DummyDeviceObject);
+
+	if (!NT_SUCCESS(status))
+	{
+		DbgPrint("Can't create root, err=%x\n", status);
+		return status;
+	}
+	return STATUS_SUCCESS;
+}
+
+static NTSTATUS NTAPI NetioComplete(
+  PDEVICE_OBJECT DeviceObject,
+  PIRP Irp,
+  PVOID Context)
+{
+	struct NetioContext *c = (struct NetioContext*) Context;
+	PIRP UserIrp = c->UserIrp;
+
+// DbgPrint("NetioComplete Irp is %p UserIrp is %p Status is 0x%08x\n", Irp, UserIrp, Irp->IoStatus.Status);
+
+	UserIrp->IoStatus.Status = Irp->IoStatus.Status;
+	UserIrp->IoStatus.Information = Irp->IoStatus.Information;
+
+	RemoveEntryList(&UserIrp->Tail.Overlay.ListEntry);
+	IoCompleteRequest(UserIrp, IO_NETWORK_INCREMENT);
+
+	SocketPut(c->socket);
+	ExFreePoolWithTag(c, 'NEIO');
+
+	return STATUS_SUCCESS;
+}
+
+static NTSTATUS WSKAPI WskControlSocket(
+    _In_ PWSK_SOCKET Socket,
+    _In_ WSK_CONTROL_SOCKET_TYPE RequestType,
+    _In_ ULONG ControlCode,
+    _In_ ULONG Level,
+    _In_ SIZE_T InputSize,
+    _In_reads_bytes_opt_(InputSize) PVOID InputBuffer,
+    _In_ SIZE_T OutputSize,
+    _Out_writes_bytes_opt_(OutputSize) PVOID  OutputBuffer,
+    _Out_opt_ SIZE_T *OutputSizeReturned,
+    _Inout_opt_ PIRP Irp)
+{
+	struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL*) Socket;
+	NTSTATUS status = STATUS_NOT_IMPLEMENTED;
+
+	if (s == NULL) {
+		DbgPrint("WskControlSocket: Socket is NULL\n");
+		status = STATUS_INVALID_PARAMETER;
+	}
+	switch (RequestType) {
+	case WskSetOption:
+		switch (Level) {
+		case SOL_SOCKET:
+			switch (ControlCode) {
+			case SO_REUSEADDR:	/* add more supported flags here ... */
+				if (InputBuffer == NULL) {
+					DbgPrint("WskControlSocket: Need an InputBuffer for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				if (InputSize < 4) {
+					DbgPrint("WskControlSocket: InputBuffer too small for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				int flag = *(int*) InputBuffer;
+				if (flag != 0) {
+					s->Flags |= ControlCode;
+				} else {
+					s->Flags &= ~ControlCode;
+				}
+				status = STATUS_SUCCESS;
+
+				break;
+			/* Windows specific. This sets the mask for callback functions: */
+			case SO_WSK_EVENT_CALLBACK:
+				if (InputBuffer == NULL) {
+					DbgPrint("WskControlSocket: Need an InputBuffer for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				if (InputSize < sizeof(WSK_EVENT_CALLBACK_CONTROL)) {
+					DbgPrint("WskControlSocket: InputBuffer too small for this operation\n");
+					status = STATUS_INVALID_PARAMETER;
+					break;
+				}
+				WSK_EVENT_CALLBACK_CONTROL *c = (WSK_EVENT_CALLBACK_CONTROL*)InputBuffer;
+
+					/* TODO: compare this bytewise */
+/*				if (c->NpiId != &NPI_WSK_INTERFACE_ID) {
+					DbgPrint("WskControlSocket: Only NPI_WSK_INTERFACE_ID supported, sorry.\n");
```

Comment:
```suggestion
					DbgPrint("WskControlSocket: Only NPI_WSK_INTERFACE_ID supported\n");
```
dots at the end of dbg-prints are not recommended. And debug-prints should be limited to relevant output. E.g. No need for apologies. This part is *optional*

Reply:
Ok dots and sorrys removed.